runtime/domain.c

```diff
@@ -373,15 +392,26 @@ asize_t caml_norm_minor_heap_size (intnat wsize)
   and if a request comes to change the minor heap size.
 
   The boundaries of this committed memory area are
-    [domain_state->young_start]
-  and
-    [domain_state->young_end].
+     [domain_state->young_start]
+   and
+     [domain_state->young_end].
```

Comment:
For a given domain, do

1. `assert (young_start == minor_heap_area)`
2. `assert (young_end == minor_heap_area + minor_heap_wsz * sizeof(value))`
3. `assert (minor_heap_area_end == caml_minor_heap_max_wsz * sizeof(value))`

always hold? 

IINM I don't think we modify `young_start` and `young_end` except at the same places where we modify `minor_heap_area` and `minor_heap_wsz`. If so,

1. Can we document these assertions here?
2. Perhaps we can get rid of `minor_heap_area` and `minor_heap_area_end` in a separate PR? They were necessary in the early versions of multicore OCaml where the domain state for a given domain was placed just before `young_start`. We did not have a dedicated register for domain state. We used the allocation pointer to find the domain state. The `minor_heap_area*` variables don't seem necessary now.

Reply:
Thanks for the perspective here.

As I documented in those comments, my mental model is that `minor_heap_area{,_end}` is about the "reserved" area for the domain minor heap, while `young_{start,end}` is about the *committed* memory for the minor heap. So no, I think that these assertions do not hold in general, `young_end` is typically sensibly smaller than `minor_heap_area_end`.

Currently `minor_heap_area{,_end}` can be recomputed easily from the domain id and `minor_heap_wsz`, because the placement scheme is very regular, so we could still do without those variables. But having them gives us an easy way to experiment with other schemes if we want to in the future. (For example, we could consider having minor heaps of different reserved sizes for each domain, or not-reusing the minor heap area of a previous domain with the same id for locality reasons.)

Finally: in the transient state where we are initializing a domain, or uninitializing a domain, or changing the minor heaps area, we can observe `young_start == young_end == NULL` while `minor_heap_area{,_end}` is already set. This is another case where the assertions/invariant you propose are not satisfied, but of course this is very temporary.

Reply:
That makes sense. Thanks!